src/Entity/SmsDeliveryReport.php

```
@@ -89,30 +96,64 @@ public function setStatusMessage($message) {
   /**
    * {@inheritdoc}
    */
+  public function getStatusTime() {
+    return $this->get('status_time')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setStatusTime($timestamp) {
+    $this->set('status_time', $timestamp);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
   public function getTimeQueued() {
-    return $this->get('time_queued')->value;
+    $queued = $this->getRevisionAtStatus(SmsMessageReportStatus::QUEUED);
+    return $queued ? $queued->getStatusTime() : NULL;
   }
 
   /**
    * {@inheritdoc}
    */
   public function setTimeQueued($time) {
```

If the method is not static, and we are setting, it makes sense to set the current object, not some other revision of the object.

setTimeDelivered,setTimeQueued should not traverse other revisions, instead it should:

- set the current report status to queued
- set the status time.

---

This makes sense if there was no previous revision at that status. But what if there is already a revision at that status? We end up with multiple revisions that have the same status, which will also be tricky when retrieving the revision that has a particular status.

Or maybe throw an exception??

---

Thats ok. The system is revisions, there can be multiple of the same status over time. The rule is: only one status at a time.

---

The helper for getXStatus should be: get most recent X status